charmhelpers/contrib/openstack/deferred_events.py

```diff
@@ -0,0 +1,291 @@
+                    event['time'],
+                    event['service'],
+                    'Pkg Update',
+                    event['action']))
+        return queued_restarts
+
+    def clear_deferred_events(self, services, action):
+        """Clear deferred events.
+
+        :param services: Services with deferred actions to clear.
+        :type services: List[str]
+        :param action: Action that was deferred
+        :type action: str
+        """
+        policy_rcd.clear_deferred_pkg_events(services, action)
+
+
+def get_deferred_events():
+    """Return a list of deferred events requested by the charm and packages.
+
+    :returns: List of deferred events
+    :rtype: List[ServiceEvent]
+    """
+    return DeferredCharmServiceEvents().events + DeferredPackageServiceEvents().events
+
+
+def get_deferred_restarts():
+    """List of deferred restart events requested by the charm and packages.
+
+    :returns: List of deferred restarts
+    :rtype: List[ServiceEvent]
+    """
+    return [e for e in get_deferred_events() if e.action == 'restart']
+
+
+def clear_deferred_events(services, action):
+    """Clear deferred events of type `action` targetted at `services`.
+
+    :param services: Services with deferred actions to clear.
+    :type services: List[str]
+    :param action: Action to clear
+    :type action: str
+    """
+    DeferredCharmServiceEvents().clear_deferred_events(services, action)
+    DeferredPackageServiceEvents().clear_deferred_events(services, action)
+
+
+def clear_deferred_restarts(services):
+    """Clear deferred restart events targetted at `services`.
+
+    :param services: Services with deferred actions to clear.
+    :type services: List[str]
+    """
+    clear_deferred_events(services, 'restart')
+
+
+def process_svc_restart(service):
+    """Respond to a service restart having occured.
+
+    :param service: Services that the action was performed against.
+    :type service: str
+    """
+    clear_deferred_restarts([service])
+
+
+def is_restart_permitted():
+    """Check whether restarts are permitted.
+
+    :returns: Whether restarts are permitted
+    :rtype: bool
+    """
+    if hookenv.config('enable-auto-restarts') is None:
+        return True
+    return hookenv.config('enable-auto-restarts')
+
+
+def defer_restart_on_changed(service, changed_files):
+    """Check if restarts are permitted, if they are not defer them.
+
+    :param service: Service to be restarted
+    :type service: str
+    :param changed_files: Files that have changed to trigger restarts.
+    :type changed_files: List[str]
+    :returns: Whether restarts are permitted
+    :rtype: bool
+    """
+    permitted = is_restart_permitted()
+    if not permitted:
+        charm_events = DeferredCharmServiceEvents()
+        charm_events.add_event(
+            service,
+            event_reason='File(s) changed: {}'.format(
+                ', '.join(changed_files)),
+            action='restart')
+    return permitted
```

Comment:
defer_restart_on_changed should probably call deferrable_svc_restart and set the reason to the changed files. This will avoid duplicate code.

Reply:
Unlike deferrable_svc_restart, defer_restart_on_changed doesn't perform any service action. So, I don't think it make sense for it to call deferrable_svc_restart

Reply:
Ack. Mabye this is a naming problem. `should_defer_restart_on_changed` perhaps. I just think it is unclear what one function is used for and what the other is.